src/modules/form/hooks/useDynamicViewFields.tsx

```diff
@@ -0,0 +1,107 @@
+import { isEqual, omit } from 'lodash-es';
+import { useCallback, useEffect, useMemo, useState } from 'react';
+
+import DynamicViewFields, {
+  isShown,
+  Props as DynamicViewFieldsProps,
+} from '../components/viewable/DynamicViewFields';
+import { addDescendants } from '../util/formUtil';
+
+import useComputedData from './useComputedData';
+
+import usePrevious from '@/hooks/usePrevious';
+import { FormDefinitionJson, FormItem } from '@/types/gqlTypes';
+
+const useDynamicViewFields = ({
+  definition,
+  values: baseValues,
+  bulk,
+}: {
+  definition: FormDefinitionJson;
+  values: Record<string, any>;
+  bulk?: boolean;
+}) => {
+  const [values, setValues] = useState<typeof baseValues>(baseValues);
+  const prevBaseValues = usePrevious(baseValues);
+
+  useEffect(() => {
+    if (!isEqual(baseValues, prevBaseValues)) setValues(baseValues);
+  }, [baseValues, prevBaseValues]);
+
+  const {
+    itemMap,
+    autofillDependencyMap,
+    enabledDependencyMap,
+    initiallyDisabledLinkIds = [],
+  } = useComputedData({ definition, initialValues: values });
+
+  const [disabledLinkIds, setDisabledLinkIds] = useState(
+    initiallyDisabledLinkIds
+  );
+
+  // Get form state, with "hidden" fields (and their children) removed
+  const getCleanedValues = useCallback(() => {
+    if (!definition) return values;
+    const excluded = addDescendants(disabledLinkIds, definition);
+    return omit(values, excluded);
+  }, [definition, disabledLinkIds, values]);
+
+  const shouldShowItem = useCallback(
+    (item: FormItem) => isShown(item, disabledLinkIds),
+    [disabledLinkIds]
+  );
+
+  const renderFields = useCallback(
+    (
+      props: Omit<
+        DynamicViewFieldsProps,
+        | 'definition'
+        | 'itemMap'
+        | 'autofillDependencyMap'
+        | 'enabledDependencyMap'
+        | 'disabledLinkIds'
+        | 'setDisabledLinkIds'
+        | 'values'
+        | 'setValues'
+        | 'bulk'
+      >
+    ) => (
+      <DynamicViewFields
+        {...props}
+        {...{
+          definition,
+          itemMap,
+          autofillDependencyMap,
+          enabledDependencyMap,
+          disabledLinkIds,
+          setDisabledLinkIds,
+          setValues,
```

Comment:
Yes that's exactly why. It has to fetch the pick list stuff, then that will cause an update which will set the diabledLinkIds correctly based on the fetched pick list values.

Reply:
What do you think of implementing [this suggestion](https://github.com/greenriver/hmis-frontend/pull/214#discussion_r1162118729) so that we can remove all these handlers? The display string change can be handled locally in the display component.

I don't think that the picklist changes could affect the disabledLinkIds, since that should always based on the "code"  and not the "label."

Reply:
See my comment above. I think they definitely do affect the disabledLinkIds currently